frame/election-provider-multi-phase/src/lib.rs

```diff
@@ -0,0 +1,1456 @@
+		type Currency: ReservableCurrency<Self::AccountId> + Currency<Self::AccountId>;
+
+		/// Duration of the unsigned phase.
+		#[pallet::constant]
+		type UnsignedPhase: Get<Self::BlockNumber>;
+		/// Duration of the signed phase.
+		#[pallet::constant]
+		type SignedPhase: Get<Self::BlockNumber>;
+
+		/// The minimum amount of improvement to the solution score that defines a solution as
+		/// "better" (in any phase).
+		#[pallet::constant]
+		type SolutionImprovementThreshold: Get<Perbill>;
+
+		/// The priority of the unsigned transaction submitted in the unsigned-phase
+		type MinerTxPriority: Get<TransactionPriority>;
+		/// Maximum number of iteration of balancing that will be executed in the embedded miner of
+		/// the pallet.
+		type MinerMaxIterations: Get<u32>;
+		/// Maximum weight that the miner should consume.
+		///
+		/// The miner will ensure that the total weight of the unsigned solution will not exceed
+		/// this values, based on [`WeightInfo::submit_unsigned`].
+		type MinerMaxWeight: Get<Weight>;
+
+		/// Something that will provide the election data.
+		type DataProvider: ElectionDataProvider<Self::AccountId, Self::BlockNumber>;
+
+		/// The compact solution type
+		type CompactSolution: codec::Codec
+			+ Default
+			+ PartialEq
+			+ Eq
+			+ Clone
+			+ sp_std::fmt::Debug
+			+ CompactSolution;
+
+		/// Accuracy used for fallback on-chain election.
+		type OnChainAccuracy: PerThing128;
+
+		/// Configuration for the fallback
+		type Fallback: Get<FallbackStrategy>;
+
+		/// The configuration of benchmarking.
+		type BenchmarkingConfig: BenchmarkingConfig;
+
+		/// The weight of the pallet.
+		type WeightInfo: WeightInfo;
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_initialize(now: T::BlockNumber) -> Weight {
+			let next_election = T::DataProvider::next_election_prediction(now).max(now);
+
+			let signed_deadline = T::SignedPhase::get() + T::UnsignedPhase::get();
+			let unsigned_deadline = T::UnsignedPhase::get();
+
+			let remaining = next_election - now;
+			let current_phase = Self::current_phase();
+
+			match current_phase {
+				Phase::Off if remaining <= signed_deadline && remaining > unsigned_deadline => {
+					Self::on_initialize_open_signed();
+					log!(info, "Starting signed phase at #{:?} , round {}.", now, Self::round());
+					T::WeightInfo::on_initialize_open_signed()
+				}
+				Phase::Signed | Phase::Off
+					if remaining <= unsigned_deadline && remaining > 0u32.into() =>
+				{
+					let (need_snapshot, enabled, additional) = if current_phase == Phase::Signed {
+						// followed by a signed phase: close the signed phase, no need for snapshot.
+						// TWO_PHASE_NOTE: later on once we have signed phase, this should return
+						// something else.
+						(false, true, Weight::zero())
+					} else {
+						// no signed phase: create a new snapshot, definitely `enable` the unsigned
+						// phase.
+						(true, true, Weight::zero())
+					};
```

Comment:
why `enabled` returned by the condition here ? it is always true.

(sidenote: also the phase 3 doesn't change this code, at least for now)

Reply:
The bool is explained here: 

https://github.com/paritytech/substrate/pull/7909/files#diff-a8382c7a11352cf83fe1e60f06e0a9059a00443155b0a8e3189de37bc9ac4fedR174-R179

But I will consider removing them. We might only want to **not enable** an unsigned phase if we have a good signed phase. But with the current planning, I think we will push forward to merge and deploy the unsigned phase asap and then add the signed one.